---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r initial setup, echo=TRUE}

# Load the required libraries
library(stringr)
library(edgeR)
library(dplyr)
library(survival)
library(ggplot2)
library(survminer)
library(clusterProfiler)
library(GSVA)
library(fgsea)
library(enrichplot)
library(qusage)
library(ggupset)
library(europepmc)

```

```{r metadata loading, echo=TRUE}

# Load metadata (ask user to provide rowname ID column name if not in the first column)
temp_metadata <- read.csv("E:/Projects/Example_data/TCGA_PAAD/TCGA.PAAD.sampleMap%2FPAAD_clinicalMatrix", header = TRUE, sep = "\t")
# Remove erratic characters from rownames
rownames(temp_metadata) <- unlist(lapply(temp_metadata[,1], function(x) str_replace_all(x, "-","_")))
# Remove erratic characters from colnames

# Check for NA values

# Add outcomes
temp_extra <- read.csv("E:/Projects/Example_data/TCGA_PAAD/survival_PAAD_survival.txt", header = TRUE, sep = "\t")
rownames(temp_extra) <- unlist(lapply(temp_extra[,1], function(x) str_replace_all(x, "-","_")))
colnames(temp_extra) <- paste0("Outcomes_",colnames(temp_extra))

temp_metadata <- temp_metadata %>% mutate(Row.names = rownames(.))
temp_extra <- temp_extra %>% mutate(Row.names = rownames(.))
temp_metadata <- temp_metadata %>% left_join(temp_extra, by = "Row.names")
rownames(temp_metadata) <- temp_metadata$Row.names
temp_metadata <- temp_metadata %>% select(-Row.names)


```

```{r RNAseq pre-processing, echo=TRUE}

# Load RNAseq data (ask user to provide rowname ID column name if not in the first column)
temp_omicdata <- read.csv("E:/Projects/Example_data/TCGA_PAAD/HiSeqV2", header = TRUE, sep = "\t")
# Check for duplicate gene names before setting to rows

# Assign genes to rownames
rownames(temp_omicdata) <- temp_omicdata[,1]
# Drop rowname column after assignment
temp_omicdata <- temp_omicdata[,-1]
# Remove erratic characters from colnames
colnames(temp_omicdata) <- unlist(lapply(colnames(temp_omicdata), function(x) str_replace_all(x, "[.]","_")))
# Check for NA values

# Check if data is in log scale and convert to counts if it is
if(max(temp_omicdata) < 1000){
  temp_omicdata <- (2^temp_omicdata)-1
}

# Convert to DGE object and filter lowly expressed genes
isexpr <- rowSums(cpm(temp_omicdata) > 0.1) >= 5
dge <- DGEList(temp_omicdata[isexpr, ])
dge <- calcNormFactors(dge)

```

```{r derive Purist subtypes}

### From https://github.com/naimurashid/PurIST/tree/master
load("E:/Projects/Example_data/fitteds_public_2019-02-12.Rdata")
classifier = classifs[[1]]
temp_classifier <- list()
temp_classifier[["TSPs"]] <- classifier$TSPs[classifier$fit$beta[-1]!=0,]
#temp_classifier[["intercept"]] <- classifier$fit$beta[1]
temp_classifier[["coefs"]] <- classifier$fit$beta[classifier$fit$beta!=0,]


purist_ind_fun = function(train_sub, input_classifier){
  indmat = matrix(-1, ncol(train_sub), nrow(input_classifier$TSPs))
  for(i in 1:nrow(input_classifier$TSPs)){
    p1 = which(rownames(train_sub) == input_classifier$TSPs[i,1])
    p2 = which(rownames(train_sub) == input_classifier$TSPs[i,2])
    if(length(p1) == 0) stop(sprintf("%s is not found in input matrix rownames",input_classifier$TSPs[i,1]))
    if(length(p2) == 0) stop(sprintf("%s is not found in input matrix rownames",input_classifier$TSPs[i,2]))
    indmat[,i] = (train_sub[p1,] > train_sub[p2,])^2
  }
  indmat = t(indmat)
  colnames(indmat) = colnames(train_sub)
  return(indmat)
}

apply_purist_classifier = function(temp_data, input_classifier){
  
  # create TSP indicator matrix
  indmat = t(purist_ind_fun(train_sub = temp_data, input_classifier = input_classifier))
  
  # name columns
  colnames(indmat) = paste("indmat", 1:ncol(indmat), sep = "")
  
  # add intercept column
  X=cbind(rep(1, nrow(indmat)), indmat)
  
  # make prediction 
  beta = input_classifier$coefs
  Pred_prob_basal = exp(X%*%beta)/(1+exp(X%*%beta))
  
  # get subtype
  Subtype = c("Classical","Basal")[(Pred_prob_basal > 0.5)^2 + 1]
  
  # get graded subtype
  Subtype_graded = rep(1, length(Pred_prob_basal))
  Subtype_graded[Pred_prob_basal < .1] = 1
  Subtype_graded[Pred_prob_basal > .1 & Pred_prob_basal < .4] = 2
  Subtype_graded[Pred_prob_basal > .4 & Pred_prob_basal < .5] = 3
  Subtype_graded[Pred_prob_basal > .5 & Pred_prob_basal < .6] = 4
  Subtype_graded[Pred_prob_basal > .6 & Pred_prob_basal < .9] = 5
  Subtype_graded[Pred_prob_basal > .9 ] = 6
  
  # graded categories
  grades = c("strong classical","likely classical","lean classical","lean basal-like","likely basal-like", "strong basal-like")
  Subtype_graded = grades[Subtype_graded]
  
  # final matrix
  final = data.frame(Pred_prob_basal= Pred_prob_basal, Subtype = Subtype, Subtype_graded = Subtype_graded)
  rownames(final) = make.names(colnames(temp_data), unique = any(table(colnames(temp_data)) > 1) )
  
  return(final)
}

# Derive TPM

# Import gene lengths to approximate TPM
temp_gene_data <- read.table("E:/Projects/Example_data/TCGA_PAAD/Full_ensembl_gene_data.txt", header = TRUE, sep = "\t")
temp_gene_data <- temp_gene_data[order(temp_gene_data$gene_length, decreasing = TRUE),]
temp_gene_data <- temp_gene_data[!duplicated(temp_gene_data$hgnc_symbol),]
rownames(temp_gene_data) <- temp_gene_data$hgnc_symbol
temp_gene_data <- temp_gene_data[rownames(temp_omicdata),]
temp_gene_data <- temp_gene_data[!is.na(temp_gene_data$gene_length),]

# Calculate CPM (counts per million)
temp_cpm <- as.data.frame(cpm(temp_omicdata))
temp_cpm <- temp_cpm[rownames(temp_gene_data),]
temp_cpm <- temp_cpm[rowSums(is.na(temp_cpm)) == 0,]
temp_gene_data <- temp_gene_data[rownames(temp_cpm),]

# Example gene lengths (in kilobases)
gene_lengths <- temp_gene_data$gene_length / 1000

# Convert CPM to TPM
temp_tpm <- temp_cpm / gene_lengths
temp_tpm <- t(t(temp_tpm) / colSums(temp_tpm)) * 1e6

### Derive Purist subtypes with CPM and TPM and compare results
purist_tpm_results <- apply_purist_classifier(temp_tpm, temp_classifier)
purist_cpm_results <- apply_purist_classifier(temp_cpm, temp_classifier)

## Compare Purist subtypes between CPM and TPM data

## Add subtypes to metadata
new_metadata <- temp_metadata
new_metadata <- new_metadata %>% mutate(Row.names = rownames(.))
new_data <- purist_tpm_results
colnames(new_data) <- paste0("Purist_",colnames(new_data))
new_data <- new_data %>% mutate(Row.names = rownames(.))
new_metadata <- new_metadata %>% left_join(new_data, by = "Row.names")
rownames(new_metadata) <- new_metadata$Row.names
new_metadata <- new_metadata %>% select(-Row.names)
temp_metadata <- new_metadata

```

```{r differential analysis, echo=TRUE}

# Match the metadata to the omics input
matched_metadata <- as.data.frame(temp_metadata[colnames(temp_omicdata),])

# Create design matrix
primary_group <- "Purist_Subtype"
temp_formula <- as.formula(paste("~0 +",primary_group))
design <- model.matrix(temp_formula, matched_metadata)
colnames(design) <- gsub(primary_group, "", colnames(design))

# Create a contrast matrix
contr.matrix <- makeContrasts(
   BasalVSClassical = Basal - Classical, 
   levels = colnames(design))

# Run voom normalization
dge_voom <- voom(dge, design, plot=TRUE)

# Apply normalized data to linear model with chosen design and contrast matrix
vfit <- lmFit(dge_voom, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)

# Compute moderated stats using empirical Bayes moderation of error
efit <- eBayes(vfit)

# Summarize differential results
summary(decideTests(efit))

# Store full differential results into a table
temp_dge_results <- topTable(efit, coef = "BasalVSClassical", number = 100000)

```

```{r outcome analysis with event curves, echo=TRUE}

new_metadata <- temp_metadata
# Identify outcome of interest with associated TTE and event variables
new_metadata$temp_outcome_status <- new_metadata$Outcomes_OS
new_metadata$temp_outcome_time <- new_metadata$Outcomes_OS.time
new_metadata$temp_primary_group <- factor(new_metadata$Purist_Subtype,levels=c("Classical","Basal"))

# Create survival model for continuous variables


# Create survival model for factor variables
# Step 1: Create a Survival Object
surv_obj <- Surv(time = new_metadata$temp_outcome_time, event = new_metadata$temp_outcome_status)

# Step 2: Fit a Cox Proportional Hazards Model
cox_model <- coxph(surv_obj ~ temp_primary_group, data = new_metadata)

# Step 3: Display the Summary (Hazard Ratios Included)
temp_summary <- summary(cox_model)

# Step 4: Create Survival Curves
# Predict survival based on treatment group
surv_fit <- survfit(surv_obj ~ temp_primary_group,new_metadata)

# Step 5: Plot Survival Curves
ggsurvplot(surv_fit, data = new_metadata, conf.int = TRUE, legend.title = "Purist Subtype",
           legend.labs = c("Classical", "Basal"), xlab = "Time", ylab = "Survival Probability",
           risk.table = TRUE, xlim=c(0,1000),surv.median.line=c("v"),break.time.by=100) + 
  ggtitle(paste0("OS in TCGA PAAD by Purist Subtype\nBasal subtype HR = ",round(temp_summary$coefficients[[2]],2)))



```

```{r enrichment analysis, echo=TRUE}

# Load geneset to test
temp_genesets <- qusage::read.gmt("E:/Projects/h.all.v2024.1.Hs.symbols.gmt")
alt_gmt <- clusterProfiler::read.gmt("E:/Projects/h.all.v2024.1.Hs.symbols.gmt")

# Derive GSVA scores
temp_logcpm <- data.matrix(log2(temp_cpm + 1))
temp_gsva_param <- gsvaParam(temp_logcpm, temp_genesets, maxDiff=TRUE)
temp_gsva <- gsva(temp_gsva_param)

temp_extra <- as.data.frame(t(temp_gsva))
colnames(temp_extra) <- paste0("GSVA_",colnames(temp_extra))
temp_metadata <- temp_metadata %>% mutate(Row.names = rownames(.))
temp_extra <- temp_extra %>% mutate(Row.names = rownames(.))
temp_metadata <- temp_metadata %>% left_join(temp_extra, by = "Row.names")
rownames(temp_metadata) <- temp_metadata$Row.names
temp_metadata <- temp_metadata %>% select(-Row.names)

# Match the metadata to the omics input
matched_metadata <- as.data.frame(temp_metadata[colnames(temp_gsva),])

primary_group <- "Purist_Subtype"
temp_formula <- as.formula(paste("~0 +",primary_group))
design <- model.matrix(temp_formula, matched_metadata)
colnames(design) <- gsub(primary_group, "", colnames(design))

# Create a contrast matrix
contr.matrix <- makeContrasts(
    BasalVSClassical = Basal - Classical, 
    levels = colnames(design))

## fit the same linear model now to the GSVA enrichment scores
fit <- lmFit(temp_gsva, design)
fit <- contrasts.fit(fit, contrasts=contr.matrix)

## estimate moderated t-statistics
fit <- eBayes(fit)

# Store full differential results into a table
temp_results <- topTable(fit, coef = "BasalVSClassical", number = 100000)

### Run enrichment using overrepresentation analysis
temp_up_genes <- rownames(temp_dge_results[temp_dge_results$adj.P.Val < 0.05 & temp_dge_results$logFC > 0,])
temp_down_genes <- rownames(temp_dge_results[temp_dge_results$adj.P.Val < 0.05 & temp_dge_results$logFC < 0,])
temp_up_results <- enricher(temp_up_genes, TERM2GENE = alt_gmt)
temp_down_results <- enricher(temp_down_genes, TERM2GENE = alt_gmt)

barplot(temp_up_results, showCategory=20)
barplot(temp_down_results, showCategory=20)

# Run enrichment using GSEA
temp_logfc <- temp_dge_results$logFC
names(temp_logfc) <- rownames(temp_dge_results)
temp_logfc <- sort(temp_logfc, decreasing = TRUE)
temp_gsea_results <- GSEA(temp_logfc, TERM2GENE = alt_gmt)

## Visualization of enrichment results

cnetplot(temp_up_results, categorySize="adj.P.Val", color.params = list(foldChange = temp_logfc))
cnetplot(temp_down_results, categorySize="adj.P.Val", color.params = list(foldChange = temp_logfc))

cnetplot(temp_gsea_results, categorySize="adj.P.Val", color.params = list(foldChange = temp_logfc))

heatplot(temp_gsea_results, foldChange=temp_logfc, showCategory=20)

temp_plot_input <- pairwise_termsim(temp_gsea_results)
treeplot(temp_plot_input, cluster.params = list(method = "average"))

upsetplot(temp_gsea_results)

ridgeplot(temp_gsea_results)

# Focus on individual geneset
gseaplot2(temp_gsea_results, geneSetID = 1, title = temp_gsea_results$Description[1])

gseaplot2(temp_gsea_results, geneSetID = 1:3, pvalue_table = TRUE,
          color = c("#E495A5", "#86B875", "#7DB0DD"), ES_geom = "dot")

# Check terms by publications
temp_terms <- unlist(lapply(temp_gsea_results@result$Description[temp_gsea_results@result$p.adjust < 0.05],
                            function(x){str_replace_all(str_replace(x,"HALLMARK_",""),"_"," ")}))
pmcplot(temp_terms[1:5], 2016:2024, proportion=FALSE)

```

```{r comparative covariate analysis, echo=TRUE}

## Create correlation between set of variables
# Start with correlation between subtypes and select GSVA scores and clinical variables

## Compare variables between selected groups


```

```{r comparison to outside multi-modal data sets, echo=TRUE}


```


```{r get all gene IDs and lengths for storage, echo=TRUE}

# Retrieve all human genes with HGNC, Ensembl, and Entrez IDs
gene_data <- getBM(
    attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene_id"),  # Fields to retrieve
    mart = ensembl
)

transcript_data <- getBM(
    attributes = c("ensembl_gene_id", "ensembl_transcript_id", "transcript_length"),  # Fields to retrieve
    mart = ensembl
)

gene_lengths <- transcript_data %>%
    group_by(ensembl_gene_id) %>%
    summarize(gene_length = mean(transcript_length, na.rm = TRUE))  # Use mean or sum

gene_data <- gene_data %>%
    left_join(gene_lengths, by = "ensembl_gene_id")

write.table(gene_data, "E:/Projects/Example_data/TCGA_PAAD/Full_ensembl_gene_data.txt", sep = "\t", quote = FALSE, row.names = FALSE)
write.table(transcript_data, "E:/Projects/Example_data/TCGA_PAAD/Full_ensembl_transcript_data.txt", sep = "\t", quote = FALSE, row.names = FALSE)


```

```{r workflow notes}

# 1.differential results within each modality (groups selected by user) DONE
# 2.comparison to outside multi-modal data sets (most relevant in cancer due to availability)
# 3.alignment and enrichment across modalities (use EnrichR genesets)
# 4.outcome analysis (survival or any included outcome of interest) DONE
# 5.comparative covariate analysis (e.g. correlations and differences between groups)

```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
